mediatek: add support for ipTIME A6004MX

koreapyj opened the pull request:
Add basic support for ipTIME A6004MX.

Hardware:
- SoC: MediaTek MT7629 Cortex-A7 (ARMv7 1.25GHz, Dual-Core)
- RAM: DDR3 128MB
- Flash: Macronix MX35LF1GE4AB (SPI-NAND 128MB)
- WiFi: MediaTek MT7761N (2.4GHz) / MediaTek MT7762N (5GHz) - no driver
- Ethernet: SoC (WAN) / MediaTek MT7531 (LAN x4)
- UART: [GND, RX, TX, 3.3V] (115200)

Installation:
- Flash recovery image with TFTP recovery

Revert to stock firmware:
- Flash stock firmware with TFTP recovery

TFTP Recovery method:
1. Unplug the router
2. Hold the reset button and plug in
3. Release when the power LED stops flashing and go off
4. Set your computer IP address manually to 192.168.0.x / 255.255.255.0
5. Flash image with TFTP client to 192.168.0.1

koreapyj commented:
mtk-scpsys failure
```
[    2.180000] mtk-scpsys 10006000.power-controller: Failed to power on domain wb
[    2.180000] ------------[ cut here ]------------
[    2.180000] WARNING: CPU: 0 PID: 1 at drivers/soc/mediatek/mtk-scpsys.c:535 scpsys_probe+0x320/0x3a8
[    2.180000] Modules linked in:
[    2.180000] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 5.10.100 #0
[    2.180000] Hardware name: Mediatek Cortex-A7 (Device Tree)
[    2.180000] [<c0108084>] (unwind_backtrace) from [<c0104cd4>] (show_stack+0x10/0x14)
[    2.180000] [<c0104cd4>] (show_stack) from [<c03f2f48>] (dump_stack+0x94/0xa8)
[    2.180000] [<c03f2f48>] (dump_stack) from [<c0117530>] (__warn+0xb8/0x114)
[    2.180000] [<c0117530>] (__warn) from [<c01175f4>] (warn_slowpath_fmt+0x68/0x78)
[    2.180000] [<c01175f4>] (warn_slowpath_fmt) from [<c045208c>] (scpsys_probe+0x320/0x3a8)
[    2.180000] [<c045208c>] (scpsys_probe) from [<c047b2e8>] (platform_drv_probe+0x48/0x9c)
[    2.180000] [<c047b2e8>] (platform_drv_probe) from [<c047955c>] (really_probe+0xfc/0x4c4)
[    2.180000] [<c047955c>] (really_probe) from [<c0479c04>] (driver_probe_device+0x5c/0xb4)
[    2.180000] [<c0479c04>] (driver_probe_device) from [<c0479ee8>] (device_driver_attach+0xa8/0xb0)
[    2.180000] [<c0479ee8>] (device_driver_attach) from [<c0479f6c>] (__driver_attach+0x7c/0x110)
[    2.180000] [<c0479f6c>] (__driver_attach) from [<c0477678>] (bus_for_each_dev+0x64/0x90)
[    2.180000] [<c0477678>] (bus_for_each_dev) from [<c0478898>] (bus_add_driver+0xf8/0x1e0)
[    2.180000] [<c0478898>] (bus_add_driver) from [<c047a578>] (driver_register+0x88/0x118)
[    2.180000] [<c047a578>] (driver_register) from [<c0101740>] (do_one_initcall+0x54/0x1e8)
[    2.180000] [<c0101740>] (do_one_initcall) from [<c0901118>] (kernel_init_freeable+0x23c/0x290)
[    2.180000] [<c0901118>] (kernel_init_freeable) from [<c07213e8>] (kernel_init+0x8/0x118)
[    2.180000] [<c07213e8>] (kernel_init) from [<c0100128>] (ret_from_fork+0x14/0x2c)
[    2.180000] Exception stack(0xc142bfb0 to 0xc142bff8)
[    2.180000] bfa0:                                     00000000 00000000 00000000 00000000
[    2.180000] bfc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    2.180000] bfe0: 00000000 00000000 00000000 00000000 00000013 00000000
[    2.180000] ---[ end trace 6e1b4d1b5e1cf83f ]---
```

koreapyj commented:
Device tree from stock firmware
```
/dts-v1/;

/ {
	#address-cells = <0x02>;
	#size-cells = <0x02>;
	compatible = "mediatek,mt7629-rfb1\0mediatek,mt7629";
	interrupt-parent = <0x01>;
	model = "MediaTek MT7629 RFB1 board";

	chosen {
		bootargs = "earlycon=uart8250,mmio32,0x11002000 console=ttyS0,115200n1 loglevel=8";
	};

	aliases {
	};

	memory {
		device_type = "memory";
		reg = <0x00 0x40000000 0x00 0x10000000>;
	};

	cpus {
		#address-cells = <0x01>;
		#size-cells = <0x00>;
		enable-method = "mediatek,leopard-smp";

		cpu@0 {
			device_type = "cpu";
			compatible = "arm,cortex-a7";
			reg = <0x00>;
			clock-frequency = <0x4a817c80>;
		};

		cpu@1 {
			device_type = "cpu";
			compatible = "arm,cortex-a7";
			reg = <0x01>;
			clock-frequency = <0x4a817c80>;
		};
	};

	reserved-memory {
		#address-cells = <0x02>;
		#size-cells = <0x02>;
		ranges;

		N9-reserved-memory@41000000 {
			compatible = "mediatek,leopard-N9-reserved";
			no-map;
			reg = <0x00 0x41000000 0x00 0x200000>;
		};
	};

	wifi_emi@0 {
		compatible = "mediatek,rebb-wifi-emi-loader";
		op_mode = <0x00>;
		board_v11 = <0x01>;
	};

	oscillator@0 {
		compatible = "fixed-clock";
		#clock-cells = <0x00>;
		clock-frequency = <0x1312d00>;
		clock-output-names = "clk20m";
		linux,phandle = <0x05>;
		phandle = <0x05>;
	};

	oscillator@1 {
		compatible = "fixed-clock";
		#clock-cells = <0x00>;
		clock-frequency = <0x2625a00>;
		clock-output-names = "clkxtal";
	};

	timer {
		compatible = "arm,armv7-timer";
		interrupt-parent = <0x02>;
		interrupts = <0x01 0x0d 0xf04 0x01 0x0e 0xf04 0x01 0x0b 0xf04 0x01 0x0a 0xf04>;
		clock-frequency = <0x1312d00>;
		arm,cpu-registers-not-fw-configured;
	};

	infracfg@10000000 {
		compatible = "mediatek,leopard-infracfg\0syscon";
		reg = <0x00 0x10000000 0x00 0x1000>;
		#clock-cells = <0x01>;
		linux,phandle = <0x03>;
		phandle = <0x03>;
	};

	pericfg@10002000 {
		compatible = "mediatek,leopard-pericfg\0syscon";
		reg = <0x00 0x10002000 0x00 0x1000>;
		#clock-cells = <0x01>;
		linux,phandle = <0x06>;
		phandle = <0x06>;
	};

	ice_debug {
		compatible = "mediatek,leopard-ice_debug\0mediatek,mt2701-ice_debug";
		clocks = <0x03 0x01>;
		clock-names = "ice_dbg";
	};

	apxgpt@10004000 {
		compatible = "mediatek,mt7629-timer\0mediatek,mt6577-timer";
		reg = <0x00 0x10004000 0x00 0x80>;
		interrupts = <0x00 0x98 0x08>;
		clocks = <0x03 0x04>;
		assigned-clocks = <0x04 0x6b>;
		assigned-clock-parents = <0x04 0x38>;
		clock-names = "clk13m";
		clock-frequency = <0x989680>;
	};

	scpsys@10006000 {
		compatible = "mediatek,leopard-scpsys\0mediatek,mt7622-scpsys";
		#power-domain-cells = <0x01>;
		reg = <0x00 0x10006000 0x00 0x1000>;
		clocks = <0x04 0x5e>;
		clock-names = "hif_sel";
		assigned-clocks = <0x04 0x5e>;
		assigned-clock-parents = <0x04 0x28>;
		infracfg = <0x03>;
		linux,phandle = <0x10>;
		phandle = <0x10>;
	};

	systimer@10009000 {
		compatible = "mediatek,mtk-systimer";
		reg = <0x00 0x10009000 0x00 0x60>;
		interrupts = <0x00 0xa9 0x04 0x00 0xaa 0x04>;
		clocks = <0x05>;
		clock-names = "clk20m";
	};

	interrupt-controller@10200a80 {
		compatible = "mediatek,mt7622-sysirq\0mediatek,mt6577-sysirq";
		interrupt-controller;
		#interrupt-cells = <0x03>;
		interrupt-parent = <0x02>;
		reg = <0x00 0x10200a80 0x00 0x20>;
		linux,phandle = <0x01>;
		phandle = <0x01>;
	};

	emi@10203000 {
		compatible = "mediatek,leopard-emi\0mediatek,mt8127-emi";
		reg = <0x00 0x10203000 0x00 0x1000>;
	};

	apmixedsys@10209000 {
		compatible = "mediatek,leopard-apmixedsys\0syscon";
		reg = <0x00 0x10209000 0x00 0x1000>;
		#clock-cells = <0x01>;
		linux,phandle = <0x16>;
		phandle = <0x16>;
	};

	rng@1020f000 {
		compatible = "mediatek,leopard-rng\0mediatek,mt7623-rng";
		reg = <0x00 0x1020f000 0x00 0x100>;
		clocks = <0x03 0x02>;
		clock-names = "rng";
	};

	topckgen@10210000 {
		compatible = "mediatek,leopard-topckgen\0syscon";
		reg = <0x00 0x10210000 0x00 0x1000>;
		#clock-cells = <0x01>;
		linux,phandle = <0x04>;
		phandle = <0x04>;
	};

	watchdog@10212000 {
		compatible = "mediatek,mt7629-wdt\0mediatek,mt6589-wdt";
		reg = <0x00 0x10212000 0x00 0x1000>;
		interrupts = <0x00 0x80 0x02>;
		#reset-cells = <0x01>;
	};

	dramc@10214000 {
		compatible = "mediatek,leopard-dramc";
		reg = <0x00 0x10214000 0x00 0x1000>;
	};

	pinctrl@10217000 {
		compatible = "mediatek,leopard-pinctrl";
		reg = <0x00 0x10217000 0x00 0x8000 0x00 0x10005000 0x00 0x1000>;
		reg-names = "base\0eint";
		gpio-controller;
		#gpio-cells = <0x02>;
		interrupt-controller;
		#interrupt-cells = <0x02>;
		interrupts = <0x00 0x99 0x04>;
		interrupt-parent = <0x02>;
		linux,phandle = <0x1b>;
		phandle = <0x1b>;

		leds-pins-0 {
			linux,phandle = <0x17>;
			phandle = <0x17>;

			mux {
				function = "led";
				groups = "ephy_leds";
			};
		};

		i2c0@0 {
			linux,phandle = <0x07>;
			phandle = <0x07>;

			pins_bus {
				groups = "i2c_0";
				function = "i2c";
				bias-disable;
				drive-strength = <0x04>;
			};
		};

		nor_flash_pins {

			pins_bus {
				groups = "nor_flash_io\0nor_flash_wp\0nor_flash_hold";
				function = "nor_flash";
			};
		};

		wifi-default {
			linux,phandle = <0x09>;
			phandle = <0x09>;

			mux-wf0_5g {
				groups = "wf0_5g";
				function = "wifi";
				drive-strength = <0x04>;
			};

			mux-wf0_2g {
				groups = "wf0_2g";
				function = "wifi";
				drive-strength = <0x04>;
			};
		};

		pcie-default {
			linux,phandle = <0x11>;
			phandle = <0x11>;

			mux-pcie_clkreq {
				groups = "pcie_clkreq";
				function = "pcie";
			};

			mux-pcie_pereset {
				groups = "pcie_pereset";
				function = "pcie";
			};

			mux-pcie_wake {
				groups = "pcie_wake";
				function = "pcie";
			};
		};

		snand_default {

			pins_bus {
				groups = "snfi\0snfi_wp\0snfi_hold";
				function = "snfi";
			};
		};

		spi_default {
			linux,phandle = <0x08>;
			phandle = <0x08>;

			mux_spi {
				groups = "spi_0";
				function = "spi";
				bias-disable;
			};
		};
	};

	interrupt-controller@10300000 {
		compatible = "arm,gic-400";
		interrupt-controller;
		#interrupt-cells = <0x03>;
		interrupt-parent = <0x02>;
		reg = <0x00 0x10310000 0x00 0x1000 0x00 0x10320000 0x00 0x1000 0x00 0x10340000 0x00 0x2000 0x00 0x10360000 0x00 0x2000>;
		linux,phandle = <0x02>;
		phandle = <0x02>;
	};

	serial@11002000 {
		compatible = "mediatek,mt7622-uart\0mediatek,mt6577-uart";
		reg = <0x00 0x11002000 0x00 0x400>;
		interrupts = <0x00 0x5b 0x08>;
		clocks = <0x04 0x51 0x06 0x0b>;
		clock-names = "baud\0bus";
		assigned-clocks = <0x04 0x49>;
		assigned-clock-parents = <0x04 0x19>;
		status = "okay";
	};

	serial@11003000 {
		compatible = "mediatek,mt7622-uart\0mediatek,mt6577-uart";
		reg = <0x00 0x11003000 0x00 0x400>;
		interrupts = <0x00 0x5c 0x08>;
		clocks = <0x04 0x51 0x06 0x0c>;
		clock-names = "baud\0bus";
		assigned-clocks = <0x04 0x49>;
		assigned-clock-parents = <0x04 0x19>;
		status = "disabled";
	};

	serial@11004000 {
		compatible = "mediatek,mt7622-uart\0mediatek,mt6577-uart";
		reg = <0x00 0x11004000 0x00 0x400>;
		interrupts = <0x00 0x5d 0x08>;
		clocks = <0x04 0x51 0x06 0x0d>;
		clock-names = "baud\0bus";
		assigned-clocks = <0x04 0x49>;
		assigned-clock-parents = <0x04 0x19>;
		status = "disabled";
	};

	pwm@11006000 {
		compatible = "mediatek,leopard-pwm";
		reg = <0x00 0x11006000 0x00 0x1000>;
		interrupts = <0x00 0x4d 0x08>;
		clocks = <0x04 0x4d 0x06 0x08 0x06 0x01>;
		clock-names = "top\0main\0pwm1";
		assigned-clocks = <0x04 0x4d>;
		assigned-clock-parents = <0x04 0x2d>;
		status = "okay";
	};

	i2c@11007000 {
		compatible = "mediatek,leopard-i2c";
		reg = <0x00 0x11007000 0x00 0x90 0x00 0x11000100 0x00 0x80>;
		interrupts = <0x00 0x54 0x08>;
		clock-div = <0x04>;
		clocks = <0x06 0x10 0x06 0x09>;
		clock-names = "main\0dma";
		assigned-clocks = <0x04 0x49>;
		assigned-clock-parents = <0x04 0x19>;
		#address-cells = <0x01>;
		#size-cells = <0x00>;
		status = "okay";
		pinctrl-names = "default";
		pinctrl-0 = <0x07>;
	};

	snfi@1100d000 {
		compatible = "mediatek,leopard-snand";
		reg = <0x00 0x1100d000 0x00 0x2000>;
		interrupts = <0x00 0x60 0x08>;
		clocks = <0x06 0x13 0x06 0x12 0x06 0x14>;
		clock-names = "nfi_clk\0pad_clk\0nfiecc_clk";
		assigned-clocks = <0x04 0x49 0x04 0x4f>;
		assigned-clock-parents = <0x04 0x19 0x04 0x2e>;
		#address-cells = <0x01>;
		#size-cells = <0x00>;
		status = "okay";

		flash@0 {

			partitions {
				compatible = "fixed-partitions";
				#address-cells = <0x01>;
				#size-cells = <0x01>;

				partition@0 {
					label = "Bootloader";
					reg = <0x00 0x100000>;
					read-only;
				};

				partition@100000 {
					label = "Config";
					reg = <0x100000 0x40000>;
				};

				partition@140000 {
					label = "Factory";
					reg = <0x140000 0x80000>;
				};

				partition@1c0000 {
					label = "KernelA";
					reg = <0x1c0000 0x2000000>;
				};

				partition@21c0000 {
					label = "KernelB";
					reg = <0x21c0000 0x2000000>;
				};

				partition@41c0000 {
					label = "User_data";
					reg = <0x41c0000 0x3400000>;
				};
			};
		};
	};

	spi@1100a000 {
		compatible = "mediatek,leopard-spi";
		#address-cells = <0x01>;
		#size-cells = <0x00>;
		reg = <0x00 0x1100a000 0x00 0x100>;
		interrupts = <0x00 0x76 0x08>;
		clocks = <0x04 0x21 0x04 0x52 0x06 0x11>;
		clock-names = "parent-clk\0sel-clk\0spi-clk";
		status = "disabled";
		pinctrl-names = "default";
		pinctrl-0 = <0x08>;

		spi@0 {
			compatible = "rohm,dh2228fv";
			reg = <0x00>;
			spi-max-frequency = <0xf4240>;
		};
	};

	spi@11014000 {
		compatible = "mediatek,mt7629-nor\0mediatek,mt8173-nor";
		reg = <0x00 0x11014000 0x00 0xe0>;
		clocks = <0x06 0x15 0x04 0x50>;
		clock-names = "spi\0sf";
		#address-cells = <0x01>;
		#size-cells = <0x00>;
		status = "okay";

		flash@0 {
			compatible = "jedec,spi-nor";
			#address-cells = <0x01>;
			#size-cells = <0x01>;

			partition@00000 {
				label = "Bootloader";
				reg = <0x00 0x60000>;
			};

			partition@60000 {
				label = "Config";
				reg = <0x60000 0x10000>;
			};

			partition@70000 {
				label = "Factory";
				reg = <0x70000 0x40000>;
			};

			partition@b0000 {
				label = "Kernel";
				reg = <0xb0000 0xb50000>;
			};
		};
	};

	wbsys@18000000 {
		compatible = "mediatek,wbsys";
		reg = <0x00 0x18000000 0x00 0x100000>;
		interrupts = <0x00 0xd3 0x08 0x00 0x81 0x08>;
		chip_id = <0x7629>;
		pinctrl-names = "default";
		pinctrl-0 = <0x09>;
		status = "okay";
	};

	ssusbsys@1a000000 {
		compatible = "mediatek,leopard-ssusbsys\0syscon";
		reg = <0x00 0x1a000000 0x00 0x1000>;
		#clock-cells = <0x01>;
		#reset-cells = <0x01>;
		linux,phandle = <0x0a>;
		phandle = <0x0a>;
	};

	usb@1a0c0000 {
		compatible = "mediatek,mt7622-xhci\0mediatek,mt2701-xhci";
		reg = <0x00 0x1a0c0000 0x00 0x1000 0x00 0x1a0c3e00 0x00 0x100>;
		reg-names = "mac\0ippc";
		interrupts = <0x00 0xe8 0x08>;
		clocks = <0x0a 0x03 0x0a 0x02 0x0a 0x04 0x0a 0x05>;
		clock-names = "sys_ck\0free_ck\0ahb_ck\0dma_ck";
		phys = <0x0b 0x03 0x0c 0x04>;
		status = "okay";
	};

	usb-phy@1a0c4000 {
		compatible = "mediatek,mt7622-u3phy\0mediatek,mt2712-u3phy";
		#address-cells = <0x02>;
		#size-cells = <0x02>;
		ranges;
		status = "okay";

		usb-phy@1a0c4000 {
			reg = <0x00 0x1a0c4000 0x00 0x700>;
			#phy-cells = <0x01>;
			clocks = <0x0a 0x01>;
			clock-names = "ref";
			linux,phandle = <0x0b>;
			phandle = <0x0b>;
		};

		usb-phy@1a0c4700 {
			reg = <0x00 0x1a0c4700 0x00 0x700>;
			#phy-cells = <0x01>;
			clocks = <0x05>;
			clock-names = "ref";
			linux,phandle = <0x0c>;
			phandle = <0x0c>;
		};
	};

	pciesys@1a100800 {
		compatible = "mediatek,leopard-pciesys\0syscon";
		reg = <0x00 0x1a100800 0x00 0x1000>;
		#clock-cells = <0x01>;
		#reset-cells = <0x01>;
		linux,phandle = <0x0d>;
		phandle = <0x0d>;
	};

	pcie@1a140000 {
		compatible = "mediatek,leopard-pcie";
		device_type = "pci";
		reg = <0x00 0x1a140000 0x00 0x1000 0x00 0x1a145000 0x00 0x1000>;
		reg-names = "subsys\0port1";
		#address-cells = <0x03>;
		#size-cells = <0x02>;
		interrupts = <0x00 0xb0 0x08 0x00 0xe5 0x08>;
		clocks = <0x0d 0x04 0x0d 0x08 0x0d 0x00 0x0d 0x03 0x0d 0x01 0x0d 0x05>;
		clock-names = "sys_ck1\0ahb_ck1\0aux_ck1\0axi_ck1\0obff_ck1\0pipe_ck1";
		assigned-clocks = <0x04 0x5f 0x04 0x49 0x04 0x5e>;
		assigned-clock-parents = <0x04 0x2d 0x04 0x19 0x04 0x28>;
		resets = <0x0e 0x03>;
		reset-names = "phy-rst1";
		phys = <0x0f 0x02>;
		phy-names = "pcie-phy1";
		power-domains = <0x10 0x01>;
		bus-range = <0x00 0xff>;
		ranges = <0x82000000 0x00 0x20000000 0x00 0x20000000 0x00 0x10000000>;
		pinctrl-names = "default";
		pinctrl-0 = <0x11>;

		pcie@1,0 {
			device_type = "pci";
			reg = <0x800 0x00 0x00 0x00 0x00>;
			#address-cells = <0x03>;
			#size-cells = <0x02>;
			#interrupt-cells = <0x01>;
			ranges;
			num-lanes = <0x01>;
			interrupt-map-mask = <0x00 0x00 0x00 0x07>;
			interrupt-map = <0x00 0x00 0x00 0x01 0x12 0x00 0x00 0x00 0x00 0x02 0x12 0x01 0x00 0x00 0x00 0x03 0x12 0x02 0x00 0x00 0x00 0x04 0x12 0x03>;

			interrupt-controller {
				interrupt-controller;
				#address-cells = <0x00>;
				#interrupt-cells = <0x01>;
				linux,phandle = <0x12>;
				phandle = <0x12>;
			};
		};
	};

	pcie-phy@1a148000 {
		compatible = "mediatek,generic-tphy-v2";
		#address-cells = <0x02>;
		#size-cells = <0x02>;
		ranges;

		port1phy {
			reg = <0x00 0x1a14a000 0x00 0x1000>;
			clocks = <0x05>;
			clock-names = "ref";
			#phy-cells = <0x01>;
			linux,phandle = <0x0f>;
			phandle = <0x0f>;
		};
	};

	phy-rst@1af01000 {
		compatible = "mediatek,leopard-pciephy\0syscon";
		reg = <0x00 0x1af01000 0x00 0x08>;
		#reset-cells = <0x01>;
		linux,phandle = <0x0e>;
		phandle = <0x0e>;
	};

	syscon@1b000000 {
		#address-cells = <0x01>;
		#size-cells = <0x01>;
		compatible = "mediatek,leopard-ethsys\0syscon";
		reg = <0x00 0x1b000000 0x00 0x1000>;
		#clock-cells = <0x01>;
		#reset-cells = <0x01>;
		linux,phandle = <0x13>;
		phandle = <0x13>;
	};

	ethernet@1b100000 {
		compatible = "mediatek,mt7629-eth\0syscon";
		reg = <0x00 0x1b100000 0x00 0x20000>;
		interrupts = <0x00 0xdf 0x08 0x00 0xe0 0x08 0x00 0xe1 0x08>;
		clocks = <0x04 0x4c 0x04 0x4e 0x13 0x04 0x13 0x03 0x13 0x02 0x13 0x01 0x13 0x00 0x14 0x00 0x14 0x01 0x14 0x02 0x14 0x03 0x15 0x00 0x15 0x01 0x15 0x02 0x15 0x03 0x16 0x05 0x16 0x04>;
		clock-names = "ethif\0sgmiitop\0esw\0gp0\0gp1\0gp2\0fe\0sgmii_tx250m\0sgmii_rx250m\0sgmii_cdr_ref\0sgmii_cdr_fb\0sgmii2_tx250m\0sgmii2_rx250m\0sgmii2_cdr_ref\0sgmii2_cdr_fb\0sgmii_ck\0eth2pll";
		assigned-clocks = <0x04 0x4c 0x04 0x4e>;
		assigned-clock-parents = <0x04 0x28 0x04 0x37>;
		power-domains = <0x10 0x00>;
		mediatek,ethsys = <0x13>;
		mediatek,sgmiisys = <0x14 0x15>;
		mediatek,infracfg = <0x03>;
		dma-coherent;
		#address-cells = <0x01>;
		#size-cells = <0x00>;
		status = "okay";
		pinctrl-names = "default";
		pinctrl-0 = <0x17>;
		linux,phandle = <0x19>;
		phandle = <0x19>;

		mac@0 {
			compatible = "mediatek,eth-mac";
			reg = <0x00>;
			phy-mode = "sgmii";

			fixed-link {
				speed = <0x3e8>;
				full-duplex;
				pause;
			};
		};

		mac@1 {
			compatible = "mediatek,eth-mac";
			reg = <0x01>;
			phy-handle = <0x18>;
		};

		mdio-bus {
			#address-cells = <0x01>;
			#size-cells = <0x00>;
			linux,phandle = <0x1a>;
			phandle = <0x1a>;

			ethernet-phy@0 {
				reg = <0x00>;
				phy-mode = "gmii";
				linux,phandle = <0x18>;
				phandle = <0x18>;
			};
		};
	};

	raeth@1b100000 {
		compatible = "mediatek,leopard-eth";
		reg = <0x00 0x1b100000 0x00 0x20000>;
		interrupts = <0x00 0xdf 0x08 0x00 0xe0 0x08 0x00 0xe1 0x08 0x00 0xdb 0x08 0x00 0xf0 0x08>;
		clocks = <0x04 0x4c 0x04 0x4e 0x04 0x6a 0x16 0x03 0x16 0x04 0x16 0x05 0x13 0x00 0x13 0x04 0x13 0x01 0x13 0x02 0x13 0x03 0x14 0x00 0x14 0x01 0x14 0x02 0x14 0x03 0x15 0x00 0x15 0x01 0x15 0x02 0x15 0x03>;
		clock-names = "ethif\0sgmiitop\0sgmii1top\0eth1pll\0eth2pll\0sgmipll\0eth\0esw\0gp2\0gp1\0gp0\0sgmii_tx250m\0sgmii_rx250m\0sgmii_cdr_ref\0sgmii_cdr_fb\0sgmii1_tx250m\0sgmii1_rx250m\0sgmii1_cdr_ref\0sgmii1_cdr_fb";
		assigned-clocks = <0x04 0x4c 0x04 0x4e>;
		assigned-clock-parents = <0x04 0x28 0x04 0x37>;
		power-domains = <0x10 0x00>;
		mediatek,ethsys = <0x13>;
		dma-coherent;
		#reset-cells = <0x01>;
		#address-cells = <0x01>;
		#size-cells = <0x00>;
		status = "disabled";
	};

	sgmiisys@1b128000 {
		compatible = "mediatek,leopard-sgmiisys_0\0syscon";
		reg = <0x00 0x1b128000 0x00 0x1000>;
		#clock-cells = <0x01>;
		mediatek,physpeed = "2500";
		linux,phandle = <0x14>;
		phandle = <0x14>;
	};

	sgmiisys@1b130000 {
		compatible = "mediatek,leopard-sgmiisys_1\0syscon";
		reg = <0x00 0x1b130000 0x00 0x1000>;
		#clock-cells = <0x01>;
		mediatek,physpeed = "2500";
		linux,phandle = <0x15>;
		phandle = <0x15>;
	};

	hnat@1b000000 {
		compatible = "mediatek,mtk-hnat_v3";
		reg = <0x00 0x1b100000 0x00 0x3000>;
		resets = <0x13 0x00>;
		reset-names = "mtketh";
		status = "okay";
		mtketh-wan = "eth1";
		mtketh-ppd = "eth0";
		ext-devices = "ra0\0rax0\0apcli0\0apclix0";
		mtketh-max-gmac = <0x02>;
	};

	clkitg {
		compatible = "simple-bus";

		bring-up {
			compatible = "mediatek,clk-bring-up";
			clocks = <0x16 0x00 0x16 0x01 0x16 0x02 0x05 0x05 0x05 0x05 0x05 0x05 0x04 0x60 0x04 0x63 0x04 0x64 0x05 0x05 0x04 0x68 0x04 0x6a 0x04 0x6b 0x05 0x05 0x03 0x03 0x03 0x04 0x03 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x06 0x09 0x06 0x0a 0x05 0x05 0x05 0x05 0x06 0x0f 0x05 0x05 0x05 0x05 0x05 0x06 0x15 0x0a 0x00 0x0a 0x01 0x0a 0x02 0x0a 0x03 0x0a 0x04 0x0a 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05 0x05>;
			clock-names = "0\01\02\03\04\05\06\07\08\09\010\011\012\013\014\015\016\017\018\019\020\021\022\023\024\025\026\027\028\029\030\031\032\033\034\035\036\037\038\039\040\041\042\043\044\045\046\047\048\049\050\051\052\053\054\055\056\057\058\059\060\061\062\063\064\065\066\067\068\069\070\071\072\073";
		};

		bring-up-pd-hif1 {
			compatible = "mediatek,scpsys-bring-up";
			power-domains = <0x10 0x02>;
		};
	};

	clkao {
		compatible = "simple-bus";
		status = "disabled";

		bring-up {
			compatible = "mediatek,clk-bring-up";
			clocks = <0x16 0x00 0x16 0x01 0x16 0x02 0x16 0x03 0x16 0x04 0x16 0x05 0x04 0x4c 0x04 0x4d 0x04 0x51 0x04 0x60 0x04 0x63 0x04 0x64 0x04 0x66 0x04 0x67 0x04 0x68 0x04 0x6a 0x04 0x6b 0x03 0x01 0x03 0x02 0x03 0x03 0x03 0x04 0x03 0x05 0x06 0x01 0x06 0x02 0x06 0x03 0x06 0x04 0x06 0x05 0x06 0x06 0x06 0x07 0x06 0x08 0x06 0x09 0x06 0x0a 0x06 0x0b 0x06 0x0c 0x06 0x0d 0x06 0x0e 0x06 0x0f 0x06 0x10 0x06 0x11 0x06 0x12 0x06 0x13 0x06 0x14 0x06 0x15 0x0a 0x00 0x0a 0x01 0x0a 0x02 0x0a 0x03 0x0a 0x04 0x0a 0x05 0x0d 0x00 0x0d 0x01 0x0d 0x02 0x0d 0x03 0x0d 0x04 0x0d 0x05 0x0d 0x06 0x0d 0x07 0x0d 0x08 0x0d 0x09 0x0d 0x0a 0x0d 0x0b 0x13 0x00 0x13 0x01 0x13 0x02 0x13 0x03 0x13 0x04 0x14 0x00 0x14 0x01 0x14 0x02 0x14 0x03 0x15 0x00 0x15 0x01 0x15 0x02 0x15 0x03>;
			clock-names = "0\01\02\03\04\05\06\07\08\09\010\011\012\013\014\015\016\017\018\019\020\021\022\023\024\025\026\027\028\029\030\031\032\033\034\035\036\037\038\039\040\041\042\043\044\045\046\047\048\049\050\051\052\053\054\055\056\057\058\059\060\061\062\063\064\065\066\067\068\069\070\071\072\073";
		};

		bring-up-pd-ethsys {
			compatible = "mediatek,scpsys-bring-up";
			power-domains = <0x10 0x00>;
		};

		bring-up-pd-hif0 {
			compatible = "mediatek,scpsys-bring-up";
			power-domains = <0x10 0x01>;
		};

		bring-up-pd-hif1 {
			compatible = "mediatek,scpsys-bring-up";
			power-domains = <0x10 0x02>;
		};
	};

	ioc {
		compatible = "mediatek,leopard-subsys-ioc";
		mediatek,ethsys = <0x13>;
		mediatek,eth = <0x19>;
		mediatek,en_eth;
	};

	ethphy {
		compatible = "mediatek,eth-fe-gphy";
		mediatek,mdio = <0x1a>;
		mediatek,infracfg = <0x03>;
		status = "okay";
	};

	gpio-keys {
		compatible = "gpio-keys";
		#address-cells = <0x01>;
		#size-cells = <0x00>;

		reset {
			label = "reset";
			gpios = <0x1b 0x3c 0x01>;
			linux,code = <0x198>;
		};

		wps {
			label = "wps";
			gpios = <0x1b 0x3a 0x01>;
			linux,code = <0x211>;
		};
	};

	gsw@0 {
		compatible = "mediatek,mt753x";
		mediatek,ethsys = <0x13>;
		#address-cells = <0x01>;
		#size-cells = <0x00>;
		mediatek,mdio = <0x1a>;
		mediatek,portmap = "llllw";
		mediatek,mdio_master_pinmux = <0x00>;
		reset-gpios = <0x1b 0x1c 0x00>;
		interrupt-parent = <0x1b>;
		interrupts = <0x06 0x04>;
		status = "okay";

		port@6 {
			compatible = "mediatek,mt753x-port";
			mediatek,ssc-on;
			reg = <0x06>;
			phy-mode = "sgmii";

			fixed-link {
				speed = <0x9c4>;
				full-duplex;
			};
		};
	};
};
```

mans0n commented:
The mt7629 subtarget might need some more love? @981213

981213 commented:
> The mt7629 subtarget might need some more love? @981213

It does, but there's nothing I can do. The mt7629 support submitted by MTK upstream only works with the upstream u-boot, not the one from their old SDK. However we can't just ask users to replace their u-boot because it's a bit risky.

The upstream kernel expects the 2nd core to spin and wait for a certain variable change, while their SDK u-boot shuts down the 2nd core and a different procedure should be used to bring that up again. I tried porting the code from SDK but failed, and the registers touched by that piece of code isn't documented in their datasheet.

mans0n commented:
@981213 Thanks for your explanation. So reviving CPU 1 won't be easy at the moment.
That reminds me that BMT is (was?) also one of the discrepancy between SDK and upstream kernel, which is a blocker for mt7623 nand support...

I guess the multi-CPU port support will get in soon by Ansuel, but mt776x wifi support will not come in the near future. I also had a problem with mtk-scpsys on mt7623, but it disappeared after I moved to kernel 5.10, so it's a separate issue...

<details>

```
[    2.897180] mtk-scpsys 10006000.power-controller: Failed to power on domain mfg
[    2.904467] ------------[ cut here ]------------
[    2.909096] WARNING: CPU: 0 PID: 1 at drivers/soc/mediatek/mtk-scpsys.c:535 scpsys_probe+0x3a4/0x3f4
[    2.918198] Modules linked in:
[    2.921244] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 5.10.46 #0
[    2.927216] Hardware name: Mediatek Cortex-A7 (Device Tree)
[    2.932775] [<c010da5c>] (unwind_backtrace) from [<c010a018>] (show_stack+0x10/0x14)
[    2.940486] [<c010a018>] (show_stack) from [<c04a7dd8>] (dump_stack+0x8c/0xa0)
[    2.947679] [<c04a7dd8>] (dump_stack) from [<c011bf0c>] (__warn+0xc0/0xd8)
[    2.954525] [<c011bf0c>] (__warn) from [<c011bf74>] (warn_slowpath_fmt+0x50/0x94)
[    2.961976] [<c011bf74>] (warn_slowpath_fmt) from [<c05133f0>] (scpsys_probe+0x3a4/0x3f4)
[    2.970120] [<c05133f0>] (scpsys_probe) from [<c059ea00>] (platform_drv_probe+0x48/0x98)
[    2.978176] [<c059ea00>] (platform_drv_probe) from [<c059ca70>] (really_probe+0xfc/0x3d4)
[    2.986317] [<c059ca70>] (really_probe) from [<c059cec0>] (driver_probe_device+0x5c/0xb4)
[    2.994456] [<c059cec0>] (driver_probe_device) from [<c059d0b8>] (device_driver_attach+0x58/0x60)
[    3.003288] [<c059d0b8>] (device_driver_attach) from [<c059d118>] (__driver_attach+0x58/0xcc)
[    3.011773] [<c059d118>] (__driver_attach) from [<c059af7c>] (bus_for_each_dev+0x68/0x8c)
[    3.019914] [<c059af7c>] (bus_for_each_dev) from [<c059bef4>] (bus_add_driver+0x1c8/0x1d8)
[    3.028141] [<c059bef4>] (bus_add_driver) from [<c059d8ec>] (driver_register+0x74/0x108)
[    3.036196] [<c059d8ec>] (driver_register) from [<c01018c4>] (do_one_initcall+0x84/0x1d8)
[    3.044338] [<c01018c4>] (do_one_initcall) from [<c0c0111c>] (kernel_init_freeable+0x23c/0x290)
[    3.052999] [<c0c0111c>] (kernel_init_freeable) from [<c0830518>] (kernel_init+0x8/0x118)
[    3.061141] [<c0830518>] (kernel_init) from [<c0100148>] (ret_from_fork+0x14/0x2c)
[    3.068671] Exception stack(0xc183dfb0 to 0xc183dff8)
[    3.073696] dfa0:                                     00000000 00000000 00000000 00000000
[    3.081835] dfc0: 00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
[    3.089972] dfe0: 00000000 00000000 00000000 00000000 00000013 00000000
[    3.096597] ---[ end trace b367f932e5b34646 ]---
```

</details>

musashino205 commented on target/linux/mediatek/image/mt7629.mk:
```diff
@@ -7,3 +7,13 @@ define Device/mediatek_mt7629-rfb
+  DEVICE_DTS := mt7629-iptime-a6004mx
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-usb3
+  SUPPORTED_DEVICES := iptime,a6004mx
```
```iptime,a6004mx``` is included in ```SUPPORTED_DEVICES``` by default, this line can be dropped. see [here](https://github.com/openwrt/openwrt/blob/5205010a546fed2227c208f6bd025b47398f4cf3/include/image.mk#L408)

koreapyj replied:
resolved

mrkiko commented:
I am interesting in this thread, so I am posting here to follow it! :)

981213 commented:
Here's the CPU bringup code from mtk lede sdk:

[mt7629-smp.tar.gz](https://github.com/openwrt/openwrt/files/8058820/mt7629-smp.tar.gz)

mans0n commented:
> I am interesting in this thread, so I am posting here to follow it! :)

FYI there's a "Subscribe" button at the right side of the page (Ctrl-F to find), so you don't need to comment just to follow the thread. :)

> Here's the CPU bringup code from mtk lede sdk (linux 4.4.198):

Thanks, indeed the sdk has a piece of code similar to that found in the upstream u-boot. Though I can't really understand what is going on there. :/

dangowrt commented:
Interesting to see MT7629 in the wild. I always thought it's mostly interesting for products which will be locked-down with secure boot, like the [starlink-wifi-gen2](https://github.com/SpaceExplorationTechnologies/starlink-wifi-gen2/blob/main/scripts/build_fip.sh#L20).

981213 commented:
Hi!

> Known issues:
> 
>     * WiFi not working
> 
>     * CPU 1 not working
> 
>     * scpsys related troubles
> 

I think we can just ignore these three issues as they are not device-specific.

>     * Second CPU port of MT7531 not working (DSA not support multiple CPU ports)

Are you sure there is a second CPU port? There's an alternative config possible: The second GMAC on mt7629 can be configured for use with the built-in GPHY, and on the board I have, this built-in GPHY is used as WAN instead of the 5th port on MT7531.

981213 commented on target/linux/mediatek/dts/mt7629-iptime-a6004mx.dts:
```diff
@@ -0,0 +1,332 @@
+	pinctrl-names = "default";
+	pinctrl-0 = <&i2c_pins>;
+	status = "okay";
+};
```
i2c isn't used, so it can be removed.

981213 commented on target/linux/mediatek/dts/mt7629-iptime-a6004mx.dts:
```diff
@@ -0,0 +1,332 @@
+
+&snand {
+	status = "okay";
+	mediatek,quad-spi;
```
this node needs an rewrite due to a recent spi-nand driver switch.

981213 replied:
Also, could you check whether Mediatek BMT is used? If your board uses u-boot v2014, BMT isn't optional in SDK and you need a mediatek,bmt-v2 line in the spi-nand node.

981213 commented on target/linux/mediatek/dts/mt7629-iptime-a6004mx.dts:
```diff
@@ -0,0 +1,332 @@
+
+&pciephy1 {
+	status = "okay";
+};
```
There's no PCIE cards connected. These two nodes can be removed.

981213 commented on target/linux/mediatek/dts/mt7629-iptime-a6004mx.dts:
```diff
@@ -0,0 +1,332 @@
+				port@4 {
+					reg = <4>;
+					label = "wan";
+				};
```
Please check which port is the wan port. If the wan is on mt7531, drop the ephy related nodes. Otherwise drop this port 4 node.

981213 commented on target/linux/mediatek/dts/mt7629-iptime-a6004mx.dts:
```diff
@@ -0,0 +1,332 @@
+	pinctrl-names = "default";
+	pinctrl-0 = <&spi_pins>;
+	status = "disabled";
+};
```
This node can be removed.

981213 reviewed (changes requested):
Hi! more comments below:

koreapyj committed:
kernel: add support for mtdsplit-fit offset
Support devices that has vendor custom header before FIT image.

Some devices has vendor custom header before FIT image. In this case mt-
d-split can not find FIT image and it results rootfs mount failure.
Please refer iptime,a6004mx device for further examples.

Signed-off-by: Yoonji Park <koreapyj@dcmys.kr>

dangowrt commented on target/linux/mediatek/dts/mt7629-iptime-a6004mx.dts:
```diff
@@ -0,0 +1,294 @@
+				};
+			};
+
+			partition@1c0000 {
```
As this is a NAND-based device, using squashfs and JFFS2 directly on the NAND is inacceptable. Please use UBI at least for `rootfs` and `rootfs_data` like it's done for other boards.

koreapyj replied:
This is done by mtdsplit-fit and Makefile for recovery image in this code. As a result JFFS2 is not used. Please refer 5e4d4357fb0de3605e9ce38efcd1bbc82c2ee7e2.

dangowrt replied:
Sorry, I missed https://github.com/openwrt/openwrt/commit/62fd9f97090d05637a283d594f2d02958fd36a80, so then it's all fine. Thank you for clarifying.

koreapyj commented:
[firmware-utils patch](https://patchwork.ozlabs.org/project/openwrt/patch/20220604090807.330700-1-koreapyj@dcmys.kr/) should merged first.

dangowrt commented on target/linux/mediatek/dts/mt7629-iptime-a6004mx.dts:
```diff
@@ -0,0 +1,294 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		phy0: ethernet-phy@0 {
```
Is this an additional Ethernet PHY physically present on the board?

koreapyj replied:
SoC has internal phy and WAN port is connected on it.

dangowrt commented:
Somehow BMT will still have to be handled so the area used for UBI is excluded from it...

koreapyj commented:
> Somehow BMT will still have to be handled so the area used for UBI is excluded from it...

Actually I have no idea about BMT. Is fixed layout resolves that?

dangowrt commented:
> > Somehow BMT will still have to be handled so the area used for UBI is excluded from it...
> 
> Actually I have no idea about BMT. Is fixed layout resolves that?

In a fixed layout you can use `mediatek,bmt-remap-range` to exclude the range use for UBI from bad block relocation.
See for example
https://github.com/openwrt/openwrt/blob/master/target/linux/mediatek/dts/mt7622-elecom-wrc-x3200gst3.dts#L291

koreapyj committed:
mediatek: add support for ipTIME A6004MX
Add basic support for ipTIME A6004MX.

Hardware:
SoC: MediaTek MT7629 Cortex-A7 (ARMv7 1.25GHz, Dual-Core)
RAM: DDR3 128MB
Flash: Macronix MX35LF1GE4AB (SPI-NAND 128MB)
WiFi: MediaTek MT7761N (2.4GHz) / MediaTek MT7762N (5GHz) - no driver
Ethernet: SoC (WAN) / MediaTek MT7531 (LAN x4)
UART: [GND, RX, TX, 3.3V] (115200)

Installation:
- Flash recovery image with TFTP recovery

Revert to stock firmware:
- Flash stock firmware with TFTP recovery

TFTP Recovery method:
1. Unplug the router
2. Hold the reset button and plug in
3. Release when the power LED stops flashing and go off
4. Set your computer IP address manually to 192.168.0.x / 255.255.255.0
5. Flash image with TFTP client to 192.168.0.1

Signed-off-by: Yoonji Park <koreapyj@dcmys.kr>

dangowrt commented:
Merged. Thank you!

olgil commented:
It was not mentioned anywhere, but would this make the D-Link 1950 compatible with openwrt ? (somebody linked this this on a forum)
